python/graphstorm/run/launch.py

```diff
@@ -0,0 +1,836 @@
+    for pid in set(pids + pids1):
+        if pid == "" or int(pid) == curr_pid:
+            continue
+        all_pids.append(int(pid))
+    all_pids.sort()
+    return all_pids
+
+
+def get_all_remote_pids(hosts, ssh_port, udf_command):
+    """Get all remote processes."""
+    remote_pids = {}
+    for _, host in enumerate(hosts):
+        ip, _ = host
+        # When creating training processes in remote machines, we may insert some arguments
+        # in the commands. We need to use regular expressions to match the modified command.
+        cmds = udf_command.split()
+        new_udf_command = " .*".join(cmds)
+        pids = get_remote_pids(ip, ssh_port, new_udf_command)
+        remote_pids[(ip, ssh_port)] = pids
+    return remote_pids
+
+
+def construct_torch_dist_launcher_cmd(
+    num_trainers: int,
+    num_nodes: int,
+    node_rank: int,
+    master_addr: str,
+    master_port: int,
+) -> str:
+    """Constructs the torch distributed launcher command.
+    Helper function.
+
+    Args:
+        num_trainers:
+        num_nodes:
+        node_rank:
+        master_addr:
+        master_port:
+
+    Returns:
+        cmd_str.
+    """
+    torch_cmd_template = (
+        "-m torch.distributed.launch "
+        "--nproc_per_node={nproc_per_node} "
+        "--nnodes={nnodes} "
+        "--node_rank={node_rank} "
+        "--master_addr={master_addr} "
+        "--master_port={master_port}"
+    )
+    return torch_cmd_template.format(
+        nproc_per_node=num_trainers,
+        nnodes=num_nodes,
+        node_rank=node_rank,
+        master_addr=master_addr,
+        master_port=master_port,
+    )
+
+
+def wrap_udf_in_torch_dist_launcher(
+    udf_command: str,
+    num_trainers: int,
+    num_nodes: int,
+    node_rank: int,
+    master_addr: str,
+    master_port: int,
+) -> str:
+    """Wraps the user-defined function (udf_command) with the torch.distributed.launch module:
+
+         "python3 -m torch.distributed.launch <TORCH DIST ARGS> run/some/trainer.py arg1 arg2
+
+    Args:
+        udf_command:
+        num_trainers:
+        num_nodes:
+        node_rank:
+        master_addr:
+        master_port:
+
+    Returns:
+
+    """
+    torch_dist_cmd = construct_torch_dist_launcher_cmd(
+        num_trainers=num_trainers,
+        num_nodes=num_nodes,
+        node_rank=node_rank,
+        master_addr=master_addr,
+        master_port=master_port,
+    )
+
+    # Get the python interpreter used right now.
+    # If we can not get it we go with the default `python3`
+    python_bin = sys.executable \
+        if sys.executable is not None and sys.executable != "" \
+        else "python3 "
+
+    # transforms the udf_command from:
+    #     path/to/dist_trainer.py arg0 arg1
+    # to:
+    #     python -m torch.distributed.launch [DIST TORCH ARGS] path/to/dist_trainer.py arg0 arg1
+    udf_command = " ".join(udf_command)
+    new_udf_command = f"{python_bin} {torch_dist_cmd} {udf_command}"
+    print(new_udf_command)
```

Comment:
do we need this print?

Reply:
This is useful for debugging. (It prints the launch scripts on each machine.

Reply:
i'm not sure if this is user friendly. when we print out tones of messages, users will ignore all messages.